File: principles-of-programming-languages/assignment-1/assignment-1/PrinciplesOfProgrammingLanguagesAssignment1/src/PrinciplesOfProgrammingLanguagesAssignment1/Oyun.java
// @ Rümeysa Ceran 

package PrinciplesOfProgrammingLanguagesAssignment1;

import java.util.ArrayList;
import java.util.Scanner;
import com.github.javafaker.Faker;

public class Oyun {
	int tur_sayisi;
	ArrayList<Sehir> sehirler = new ArrayList<Sehir>();
	static Scanner scanner = new Scanner(System.in);

	public Oyun() {
	};

	Faker faker = new Faker();

	// tur sayısı alma fonksiyonu
	int turSayisiniAl() {

		do {
			System.out.printf("tur sayisi: ");
		} while (!(scanner.hasNextInt()));

		int tur_sayisi = scanner.nextInt();

		return tur_sayisi;
	}

	// bir tur atma fonksiyonu
	void turAt(ArrayList<Sehir> sehirler) {
		for (int s = 0; s < sehirler.size(); s++) {
			int nufus_birler_basamagı = sehirler.get(s).nufus % 10;
			int nufus_onlar_basamagı = sehirler.get(s).nufus / 10;
			if (nufus_onlar_basamagı > 9) {
				nufus_onlar_basamagı = nufus_onlar_basamagı % 10;
			}
			int mahalle_nufus_artis_kati = nufus_birler_basamagı + nufus_onlar_basamagı;
			if (mahalle_nufus_artis_kati != 0) // birler ve onlar basamagi toplami 0 degil ise
			{
				ArrayList<Ilce> ilceler = sehirler.get(s).ilceler;
				for (int i = 0; i < ilceler.size(); i++) {
					ArrayList<Mahalle> mahalleler = ilceler.get(i).mahalleler;
					for (int m = 0; m < mahalleler.size(); m++) { // dongu ilcedeki her bir mahalle icin doner
						int olusturulacakKisiSayisi = mahalleler.get(m).nufus;
						mahalleler.get(m).nufus = mahalleler.get(m).nufus * mahalle_nufus_artis_kati; // mahalle nufusu
																										// artirilir
						olusturulacakKisiSayisi = mahalleler.get(m).nufus - olusturulacakKisiSayisi; // yeni nufusa gore
																										// olusturulmasi
																										// gereken kisi
																										// sayisi
																										// hesaplanir
						ArrayList<Kisi> yeniOlusturulanKisiler = NesneleriOlustur
								.kisileriOlustur(olusturulacakKisiSayisi); // yeni kisileri olusturur
						mahalleler.get(m).kisiler.addAll(yeniOlusturulanKisiler);
						ArrayList<Kisi> kisiler = mahalleler.get(m).kisiler;
						for (int k = 0; k < kisiler.size(); k++) { // tur sonu her kisinin yası bir artirilir
							kisiler.get(k).yas++;
						}
					}
					ilceler.get(i).nufus = (mahalleler.size()) * (mahalleler.get(0).nufus); // ilcenin yeni nufusu
																							// guncellenir
				}
				sehirler.get(s).nufus = (ilceler.size()) * (ilceler.get(0).nufus); // sehrin yeni nufusu guncellenir
			} else { // birler ve onlar basamagı toplami 0 ise
				ArrayList<Ilce> ilceler = sehirler.get(s).ilceler;
				for (int i = 0; i < ilceler.size(); i++) {
					ArrayList<Mahalle> mahalleler = ilceler.get(i).mahalleler;
					for (int m = 0; m < mahalleler.size(); m++) {
						mahalleler.get(m).nufus = mahalleler.get(m).nufus + 1; // mahalle nufusu 1 artirilir
						ArrayList<Kisi> yeniOlusturulanKisiler = NesneleriOlustur.kisileriOlustur(1); // 1 yeni kisi
																										// olusturulur
						mahalleler.get(m).kisiler.addAll(yeniOlusturulanKisiler); // yeni kisi mahalleye eklenir
						ArrayList<Kisi> kisiler = mahalleler.get(m).kisiler;
						for (int k = 0; k < kisiler.size(); k++) {
							kisiler.get(k).yas++; // kisilerin yaslar bir artirilir
						}
					}
					ilceler.get(i).nufus = (mahalleler.size()) * (mahalleler.get(0).nufus); // ilcenin yeni nufusu
																							// guncellenir
				}
				sehirler.get(s).nufus = (ilceler.size()) * (ilceler.get(0).nufus); // sehrin yeni nufusu guncellenir
			}
		}
	}

	// ekranda tur sonu goruntuleri yazdirma fonksiyonu
	void ekrandaGoster(ArrayList<Sehir> sehirler) {
		int sehir_sayisi = sehirler.size();
		ArrayList<String> ekran = new ArrayList<>();
		for (int sayac = 0; sayac < sehir_sayisi; sayac++) {

			String sehir_gosterimi = "[" + sehirler.get(sayac).nufus + "]";
			if (sayac + 1 != sehir_sayisi) {
				sehir_gosterimi = sehir_gosterimi + "-";
			}
			ekran.add(sehir_gosterimi);
		}

		int sayac = 0;
		String s = "";
		for (int i = 0; i < ekran.size(); i++) {
			if (sayac % 5 == 0) {
				s += "\n";
			}
			s += ekran.get(i);
			sayac++;
		}
		System.out.printf(s + "\n");
	}

	// ekrani full temizler
	public static void ekranTemizle() {
		try {
			if (System.getProperty("os.name").contains("Windows")) {
				new ProcessBuilder("cmd", "/c", "cls").inheritIO().start().waitFor();
			} else {
				new ProcessBuilder("clear").inheritIO().start().waitFor();
			}
		} catch (Exception e) {
			e.printStackTrace();
		}
	}

	// tum turlar sona erince kullanıcıdan alınan satir sutun bilgisiyle ilin
	// bilgilerini gosterme fonksiyonu
	void bilgileriGoster() {

		int sutun = 5;
		int toplam = this.sehirler.size();
		int satir = (int) Math.ceil(toplam / 5.0);

		int array_satir = 0;
		int array_sutun = 0;
		boolean durum;

		// Kullanıcıdan giriş alma
		do {
			durum = false;
			System.out.print("Sirasiyla satir ve sutun giriniz(0 dahil): ");

			if (scanner.hasNextInt()) {
				array_satir = scanner.nextInt();
			} else {
				durum = true;
				scanner.next(); // hatalı input temizle
				continue;
			}

			if (scanner.hasNextInt()) {
				array_sutun = scanner.nextInt();
			} else {
				durum = true;
				scanner.next();
				continue;
			}

			// sınır kontrolü
			if (array_satir < 0 || array_satir >= satir ||
					array_sutun < 0 || array_sutun >= sutun) {
				durum = true;
				System.out.println("Gecersiz deger girdiniz!");
				continue;
			}

			// index hesapla
			int index = array_satir * sutun + array_sutun;

			if (index < 0 || index >= this.sehirler.size()) {
				System.out.println("Bu konumda sehir yok!");
				durum = true;
				continue;
			}

		} while (durum);

		// Sehir bul
		int index = array_satir * sutun + array_sutun;
		Sehir sehir = this.sehirler.get(index);

		// Yazdırma
		System.out.println("Sehir: " + sehir.ad + " - Nufus: " + sehir.nufus);

		for (Ilce ilce : sehir.ilceler) {
			System.out.printf("  Ilce: %s - Nufus: %d\n", ilce.ad, ilce.nufus);

			for (Mahalle mahalle : ilce.mahalleler) {
				System.out.printf("    Mahalle: %s - Nufus: %d\n", mahalle.ad, mahalle.nufus);

				for (Kisi kisi : mahalle.kisiler) {
					System.out.println("      " + kisi.id + " - " + kisi.ad_soyad + " - " + kisi.yas);
				}
			}
		}
	}

	// tüm oyunun calıstıgı fonksiyon
	void oyunaBasla() {
		this.tur_sayisi = turSayisiniAl();
		scanner.nextLine();
		this.sehirler = NesneleriOlustur.sehirleriOlustur(this.sehirler);
		for (int i = 0; i < this.tur_sayisi; i++) {
			int kacıncıTurda = i + 1;
			turAt(this.sehirler);
			ekranTemizle();
			System.out.println("-------------- " + kacıncıTurda + ". " + "TUR SONU " + "-----------------------");
			ekrandaGoster(this.sehirler);
			BasamakKontrol.nufusBasamakSayisiKontrolu(this.sehirler);
			System.out.println("-------- " + kacıncıTurda + ". " + "TUR SONU YENI SEHIRLER ILE " + "-----------");
			ekrandaGoster(this.sehirler);
			System.out.println("-----------------------------------------------------------------------------");
		}
		bilgileriGoster();

	}

}

File: principles-of-programming-languages/assignment-2/assignment-2/PrinciplesOfProgrammingLanguagesAssignment2/src/Oyun.c
// @ Rümeysa Ceran

#include <stdbool.h>
#include "stdio.h"
#include "stdlib.h"


#include "Sehir.h"
#include "NesneleriOlustur.h"
#include "BasamakKontrol.h"

#include "Oyun.h"

Oyun OyunOlusturma(){
    Oyun this;
    this = (Oyun)malloc(sizeof(struct OYUN));
    this->turSayisiniAl = &TurSayisiniAl;
    this->turAt = &TurAt;
    this->ekrandaGoster = &EkrandaGoster;
    this->ekranTemizle = &EkranTemizle;
    this->bilgileriGoster = &BilgileriGoster;
    this->oyunaBasla = &OyunaBasla;
    this->yoket = &Yoket;
    this->sehirSayisi = malloc(sizeof(int));
    return this;
};
int TurSayisiniAl(){ //kullanicidan tur sayisi alir
   
    int tur_sayisi;

    while (1) {

        printf("tur sayisi: ");

        if (scanf("%d", &tur_sayisi) == 1) {

            int c;
            while ((c = getchar()) != '\n' && c != EOF); 
            break;
        }

        while(getchar() != '\n');
    }

    return tur_sayisi;	
}

// bir turdaki tum olayları yaptıran fonksiyondur
void TurAt(const Oyun this, Sehir* sehirler, int *sehirSayisi, const NesneleriOlustur nesneleriOlustur){

    for(int s=0; s<*sehirSayisi; s++) {

        int nufus = sehirler[s]->super->nufus;

        int birler = nufus % 10;
        int onlar = nufus / 10;
        if(onlar>9)
		{
				onlar = onlar%10;
		}
        int mahalle_nufus_artis_kati = birler + onlar; // bir mahaledeki nufus artıs kati hesaplanir
        if(mahalle_nufus_artis_kati!=0){
            Ilce* ilceler = sehirler[s]->ilceler;
            int ilceSayisi = sehirler[s]->ilceSayisi;
            for(int i=0; i<ilceSayisi; i++){
                Mahalle* mahalleler = ilceler[i]->mahalleler;
                int mahalleSayisi = ilceler[i]->mahalleSayisi;
                for(int m=0; m<mahalleSayisi; m++) {
                    int eski = mahalleler[m]->super->nufus;
                    int olusturulacakKisiSayisi = mahalleler[m]->super->nufus; 
                    mahalleler[m]->super->nufus = mahalleler[m]->super->nufus*mahalle_nufus_artis_kati; //mahalle nufusu artirilir
                    olusturulacakKisiSayisi = mahalleler[m]->super->nufus - olusturulacakKisiSayisi; // yeni nufusa gore olusturulmasi gereken kisi sayisi hesaplanir
                    Kisi* yeni = nesneleriOlustur->kisileriOlustur(nesneleriOlustur,olusturulacakKisiSayisi);//yeni kisileri olusturur

                    //mahallelerin kisilerinin boyutu artıtılır ki yeni kisiler eklenebilsin
                    Kisi* temp = realloc(mahalleler[m]->kisiler,(eski + olusturulacakKisiSayisi) * sizeof(Kisi));
                    mahalleler[m]->kisiler = temp;
                    for(int k=0; k<olusturulacakKisiSayisi; k++) {
                    mahalleler[m]->kisiler[eski + k] = yeni[k];
                    }
                    free(yeni);
                    mahalleler[m]->super->nufus = eski + olusturulacakKisiSayisi;//yeni nufus guncellenir

                    for(int k=0; k<mahalleler[m]->super->nufus; k++) {
                        mahalleler[m]->kisiler[k]->yas++;
                    }
                }
                ilceler[i]->super->nufus = (mahalleSayisi)*(mahalleler[0]->super->nufus);//ilceninde nufusu artirilir
            
            }
            sehirler[s]->super->nufus = (ilceSayisi)*(ilceler[0]->super->nufus);//en son sehrin nfusu da artirilir
        }
        else {	//birler ve onlar basamagı toplami 0 ise 
				Ilce* ilceler = sehirler[s]->ilceler;
                int ilceSayisi = sehirler[s]->ilceSayisi;
				for(int i=0; i<ilceSayisi; i++) {
					Mahalle* mahalleler = ilceler[i]->mahalleler;
                    int mahalleSayisi = ilceler[i]->mahalleSayisi;
                    for(int m=0; m<mahalleSayisi; m++) {
                        int eski = mahalleler[m]->super->nufus;
						mahalleler[m]->super->nufus = mahalleler[m]->super->nufus + 1; //mahalle nufusu 1 artirilir
						Kisi* yeni = nesneleriOlustur->kisileriOlustur(nesneleriOlustur,1); //1 yeni kisi olusturulur

                        //mahallelerin kisilerinin boyutu artıtılır ki yeni kisi eklenebilsin
                        Kisi* temp = realloc(mahalleler[m]->kisiler,(eski + 1) * sizeof(Kisi));
                        mahalleler[m]->kisiler = temp;
                        for(int k=0; k<1; k++) {
                        mahalleler[m]->kisiler[eski + k] = yeni[k];
                        }

                        free(yeni);

                        mahalleler[m]->super->nufus = eski + 1;//yeni nufus guncellenir

                        for(int k=0; k<mahalleler[m]->super->nufus; k++) {
                            mahalleler[m]->kisiler[k]->yas++;
                        }
					}		
					ilceler[i]->super->nufus = (mahalleSayisi)*(mahalleler[0]->super->nufus); //ilcenin yeni nufusu guncellenir
				}
				sehirler[s]->super->nufus = (ilceSayisi)*(ilceler[0]->super->nufus); // sehrin yeni nufusu guncellenir
		}
    }

}

//ekranda her satırda 5 il olacak sekilde gosterme fonksiyonudur
void EkrandaGoster(Sehir* sehirler, int *sehirSayisi){

    if (!sehirler || *sehirSayisi <= 0) return;

    for(int i=0; i<*sehirSayisi; i++) {

        if (!sehirler[i]) continue;

        printf("[%d]", sehirler[i]->super->nufus);

        if(i != *sehirSayisi - 1)
            printf("-");

        if((i+1) % 5 == 0)
            printf("\n");
    }

    printf("\n");
}

//ekranı full temizler
void EkranTemizle(){
    #ifdef _WIN32
        system("cls");
    #else
        system("clear");
    #endif
};

//en son kullanıcıdan sati ve sutun bilgisi alarak o sehrin bilgilerini gosterir
void BilgileriGoster(const Oyun this){

    int sutun = 5;
    int toplam = *(this->sehirSayisi);
    int satir = (toplam + sutun - 1) / sutun;

    int array_satir = 0;
    int array_sutun = 0;
    int durum;

    // satir alinir
    do {

        durum = 0;

        printf("Satir giriniz(0 dahil): ");

        if (scanf("%d", &array_satir) != 1) {

            durum = 1;

            while(getchar() != '\n');

            continue;
        }

        while(getchar() != '\n');

        if (array_satir < 0 || array_satir >= satir) {

            printf("Gecersiz satir!\n");

            durum = 1;
        }

    } while (durum);

    //sutun alinir
    do {

        durum = 0;

        printf("Sutun giriniz(0 dahil): ");

        if (scanf("%d", &array_sutun) != 1) {

            durum = 1;

            while(getchar() != '\n');

            continue;
        }

        while(getchar() != '\n');

        if (array_sutun < 0 || array_sutun >= sutun) {

            printf("Gecersiz sutun!\n");

            durum = 1;
        }

        int index = array_satir * sutun + array_sutun;

        if (index < 0 || index >= *(this->sehirSayisi)) {

            printf("Bu konumda sehir yok!\n");

            durum = 1;
        }

    } while (durum);

    int index = array_satir * sutun + array_sutun;

    Sehir sehir = this->sehirler[index];

    printf("Sehir: %s - Nufus: %d\n",
           sehir->super->ad,
           sehir->super->nufus);

    for (int i = 0; i < sehir->ilceSayisi; i++) {

        Ilce ilce = sehir->ilceler[i];

        printf("  Ilce: %s - Nufus: %d\n",
               ilce->super->ad,
               ilce->super->nufus);

        for (int j = 0; j < ilce->mahalleSayisi; j++) {

            Mahalle mahalle = ilce->mahalleler[j];

            printf("    Mahalle: %s - Nufus: %d\n",
                   mahalle->super->ad,
                   mahalle->super->nufus);

            for (int k = 0; k < mahalle->super->nufus; k++) {

                Kisi kisi = mahalle->kisiler[k];

                printf("      %d - %s - %d\n",
                       kisi->id,
                       kisi->ad_soyad,
                       kisi->yas);
            }
        }
    }
}

//tum oyunun oynandıgı fonksiyondur
void OyunaBasla(const Oyun this){

    if (!this) return;

    printf("OyunaBasla girdi\n");

    this->turSayisi = this->turSayisiniAl();// tur sayısı kullanıcıdan alinir

    //oyun boyunca kullanılıcak olan nesnealeriOlusturma ve basamakKontrolu nesneleri olusuturulur
    NesneleriOlustur nesneleriOlustur = NesneleriOlusturOlusturma();
    BasamakKontrol basamakKontrol  = BasamakKontrolOlusturma();

    //sehirler olusturulur
    this->sehirler = nesneleriOlustur->sehirleriOlustur(nesneleriOlustur);

    if (!this->sehirler) {
        printf("sehirler NULL!\n");
        return;
    }

    //oyundaki sehir sayisi guncellenir
    *(this->sehirSayisi) = nesneleriOlustur->sehirSayisi;

    //alınan tur sayısı kadar tur doner
    for(int i=0; i<this->turSayisi; i++) {

        this->turAt(this, this->sehirler, this->sehirSayisi, nesneleriOlustur);

        //tur sonu ekran temizlenir
        this->ekranTemizle();

        printf("TUR %d SON\n", i+1);

        //tur bitimi basamak kontrolu ile yeni sehirler olusturulur
        basamakKontrol->nufusBasamakSayisiKontrolu(
            basamakKontrol,
            &this->sehirler,
            this->sehirSayisi,
            nesneleriOlustur
        );

        //ve yeni olusan sehirler ile birlikte tum sehirler ekranda gosterilir
        this->ekrandaGoster(this->sehirler, this->sehirSayisi);

        if (!this->sehirler) {
            printf("BasamakKontrol sonrası sehirler NULL!\n");
            return;
        }
    }

    //turlar bittikten sonra bilgiler gosterilir
    this->bilgileriGoster(this);

    //oyun bittigi icin bellek bosaltılıt
    nesneleriOlustur->yoket(nesneleriOlustur);
    basamakKontrol->yoket(basamakKontrol);
}

//yok edici fonksiyon
void Yoket(const Oyun this){
    if(this==NULL)return;
    for(int i=0;i<*(this->sehirSayisi);i++){
        this->sehirler[i]->yoket(this->sehirler[i]);
    }
    free(this->sehirler);
    free(this);
};